### DISCUSSION TITLE
RichTextFX Q&A

### DISCUSSION THREAD

**Discussion by Jugen** [FXMisc/RichTextFX#1199] [Q&A]
Ask questions about RichTextFX here :-)

**Comment by InternetUser70** [comment]
In CodeArea - I like to select multiple lines and press TAB (or Shift-TAB) to move all lines one tab position to the right (or to the left). I think it currently is not possible. Correct ?
Is it a plan to insert that feature ?

**Reply by Symeon94** [reply]
Hello,

What do you mean by "it is currently not possible"? Can't you:
1. Capture the TAB event
2. Insert TAB at the start of each line being selected

I would expect that this is something application dependant.

**Reply by InternetUser70** [reply]
Ok, thanks. I will implement it like this

**Reply by Symeon94** [reply]
If there is some real impossibility I'm interested to know (I actually intend to implement the same in my own application in the coming weeks/days).

**Reply by InternetUser70** [reply]
I have troubles in getting the TAB pressed event. I always got it after the TAB pressed already is consumed. I always get range length = 0 and I see in the editor that my selection is removed and the new caret position is in the selected range.

I can't find a ways to capture the event before it's consumed

```
		editorPane.setOnKeyPressed(new EventHandler<KeyEvent>() {
		  @Override
		  public void handle(KeyEvent event) {
				L.info("setOnKeyPressed - key code: " + event.getCode().toString());
		  	if (event.getCode().equals(KeyCode.TAB)) {
  				L.info("setOnKeyPressed - tab pressed");
	  			IndexRange range = editorPane.getSelection();
  				L.info("setOnKeyPressed - range: " + range.toString());

```

**Reply by Symeon94** [reply]
This is a JavaFX question. In JavaFX there are two types of event catching mechanism:
* Filter
* Handler

You can get some information about it in here: https://docs.oracle.com/javafx/2/events/processing.htm
To make it simple: Events go from the window to your component in the first phase and then event flow back from the source to the event. Until someone calls `consume()` on the event, it will continue to flow through.

The **filter** happens in the first phase (the *capturing phase*) while the **handler** happens in the second phase (the *bubbling phase*).

If you want to catch the TAB key before it is taken by the text area, you need to add a filter (with `addEventFilter(KeyEvent.KEY_PRESSED, event -> {/*...*/ event.consume();})` and to consume the event. The example you gave you used a handler.

I hope this helps.

**Reply by Jugen** [reply]
In addition to @Symeon94 reply also have a look at [OverrideBehaviorDemo](https://github.com/FXMisc/RichTextFX/blob/master/richtextfx-demos/src/main/java/org/fxmisc/richtext/demo/OverrideBehaviorDemo.java)

**Reply by InternetUser70** [reply]
I tried this before as well, but won't work. Actually I set consume() as the first statement in the event handler. I expect that the key is consumed and not be used by the CodeArea. But any typing in the CodeArea will be still used.

```
		editorPane = new CodeArea();
		//editorPane.setParagraphGraphicFactory(LineNumberFactory.get(editorPane));
		editorPane.addEventHandler(KeyEvent.KEY_PRESSED , new EventHandler<KeyEvent>() {
		  @Override
		  public void handle(KeyEvent event) {
		  	event.consume();
				L.info("setOnKeyPressed - key code: " + event.getCode().toString());
		  	if (event.getCode().equals(KeyCode.TAB)) {
  				L.info("setOnKeyPressed - tab pressed");
	  			IndexRange range = editorPane.getSelection();
  				L.info("setOnKeyPressed - range: " + range.toString());
	  			if (range.getLength() > 0) {
	  				String text = editorPane.getText();
	  				L.info("setOnKeyPressed - last index: " + text.lastIndexOf("\n", range.getStart()));
	  			}
		  	}
		  }
		});

```

**Reply by Symeon94** [reply]
@InternetUser70 I refer you back to my response, you should replace your **handler** by a **filter** if you want to capture the events before they are handled by the `CodeArea`.

**Reply by InternetUser70** [reply]
Ok, yes. I was to focused on event handlers, With event filters it works. Thanks
